tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/ManagedIdentityTests.cs

```diff
@@ -0,0 +1,189 @@
+{
+    [TestClass]
+    public class ManagedIdentityTests
+    {
+        private static readonly string[] s_msi_scopes = { "https://management.azure.com" };
+        private static readonly string s_baseURL = "https://service.msidlab.com/";
+        private static readonly string s_clientId = "client_id";
+        private static Dictionary<string, string> s_envVariables = new Dictionary<string, string>();
+        private const string UserAssignedClientID = "3b57c42c-3201-4295-ae27-d6baec5b7027";
+        private const string Mi_res_id = "/subscriptions/c1686c51-b717-4fe0-9af3-24a20a41fb0c/resourcegroups/MSAL_MSI/providers/Microsoft.ManagedIdentity/userAssignedIdentities/MSAL_MSI_USERID";
+
+        public enum AzureResource
+        {
+            webapp,
+            function,
+            vm,
+            azurearc,
+            cloudshell,
+            servicefabric
+        }
+
+        [TestMethod]
+        public async Task ManagedIdentitySourceCheckAsync()
+        {
+            //Arrange
+            string result = string.Empty;
+            string exception = string.Empty;
+            string expectedClientException = "Authentication with managed identity is unavailable. " +
+                "No managed identity endpoint found.";
+
+            IConfidentialClientApplication cca = CreateCCAWithProxy(s_baseURL);
+
+            //Act
+            try
+            {
+                AuthenticationResult authenticationResult = await cca.AcquireTokenForClient(s_msi_scopes)
+                    .WithManagedIdentity()
+                    .ExecuteAsync()
+                    .ConfigureAwait(false);
+            }
+            catch (MsalClientException ex)
+            {
+                result = ex.Message;
+            }
+            catch (Exception ex)
+            {
+                exception = ex.Message;
+            }
+
+            //Assert
+            Assert.IsTrue(string.IsNullOrEmpty(exception));
+            Assert.AreSame(result, expectedClientException);
+        }
+
+        [DataTestMethod]
+        [DataRow(AzureResource.webapp, "", DisplayName = "System Identity Web App")]
+        [DataRow(AzureResource.function, "", DisplayName = "System Identity Function App")]
+        [DataRow(AzureResource.webapp, UserAssignedClientID, DisplayName = "User Identity Web App")]
+        [DataRow(AzureResource.function, UserAssignedClientID, DisplayName = "User Identity Function App")]
+        [DataRow(AzureResource.webapp, Mi_res_id, DisplayName = "ResourceID Web App")]
+        [DataRow(AzureResource.function, Mi_res_id, DisplayName = "ResourceID Function App")]
+        public async Task AcquireMSITokenAsync(AzureResource azureResource, string userIdentity)
+        {
+            //Arrange
+            AuthenticationResult result = null;
+
+            //Set the Environment Variables
+            bool isEnvironmentVariableSet = await SetEnvironmentVariablesAsync(azureResource)
```

Comment:
I think what you are doing here is "downloading" all the env variables from the service onto the box. The name should suggest that, e.g. `SetRemoveEnvVariablesAsync` or smth like that.